cpp/src/bitmask/null_mask.cu

```diff
@@ -379,6 +474,129 @@ cudf::size_type count_unset_bits(bitmask_type const *bitmask, size_type start,
   return (num_bits - detail::count_set_bits(bitmask, start, stop, stream));
 }
 
+std::vector<size_type>
+segmented_count_set_bits(bitmask_type const* bitmask,
+                         std::vector<size_type> const& indices,
+                         cudaStream_t stream) {
+  CUDF_EXPECTS(indices.size() % 2 == 0, "Invalid ranges.");
+  for (size_t i = 0; i < indices.size() / 2; i++) {
+    auto begin = indices[i * 2];
+    auto end = indices[i * 2 + 1];
+    CUDF_EXPECTS(begin >= 0, "Invalid range.");
+    CUDF_EXPECTS(end >= begin, "Invalid range.");
+  }
+
+  if (indices.size() == 0) {
+    return std::vector<size_type>{};
+  }
+  else if (bitmask == nullptr) {
+    std::vector<size_type> ret(indices.size() / 2);
+    for (size_t i = 0; i < indices.size() / 2; i++) {
+      ret[i] = indices[2 * i + 1] - indices[2 * i];
+    }
+    return ret;
+  }
+
+  size_type num_ranges = indices.size() / 2;
+  thrust::host_vector<size_type> h_first_indices(num_ranges);
+  thrust::host_vector<size_type> h_last_indices(num_ranges);
+  thrust::stable_partition_copy(
+    thrust::seq, std::begin(indices), std::end(indices),
+    thrust::make_counting_iterator(0),
+    h_first_indices.begin(), h_last_indices.begin(),
+    [](auto i) { return (i % 2) == 0; }
+  );
+
+  rmm::device_vector<size_type> d_first_indices = h_first_indices;
+  rmm::device_vector<size_type> d_last_indices = h_last_indices;
+  rmm::device_vector<size_type> d_null_counts(num_ranges, 0);
+
+  auto word_elements =
```

Comment:
I find this name confusing. Word elements is not used anywhere else. IMO something like word_set_bits or count_set_bits is much more clear.